### PR TITLE
Implemented the logic for the Create Competition Page.

### PR DESCRIPTION
**Pull request by Phoebe-Khokgawe**
The Create Competition page stored and sent the information that is received from the user to the Competitions Servlet ("/competitionList") as a json object. 
The information sent to backend is 
- userID that created the competition (hard coded for now)
- competition name
- startDate
- endDate
- participants list

Competitions Servlet:
- Inserted the competition information to the competition database.
- Go through the participants list and checked if the user exits. If the user doesn't exist insert the user into the database using the given email and name as null. 
- Add users in the participants list to the participants database.

### CODE REVIEW

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 1]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    return -1;
+  }
+
+  /**
+   * Add the given statement to SQL database.
+   *
+   * @param statement -- statement to add to SQL database.
+   * @throws SQLException
+   */
+  private void addStatementToDataBase(String statement, Connection conn) throws SQLException {
```
[the remark this answers is not in the document]
Comment (roonil-wazlib) [reply project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:176]: I also don't understand why you modularised out a function for executing prepared statements but most of the statement executions within the main code

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 2]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      try (Connection conn = pool.getConnection()) {
+
+        //add competition to database
+        addStatementToDataBase(competitionStmt, conn);
+        int competitionId = getLatestInsertedID(conn);
+
+        //add participants to database
+        for (int i = 0; i < participants.length(); i++) {
+          String email = participants.get(i).toString();
+          String findUserStmt = "SELECT name,id FROM users WHERE email=\"" + email + "\";";
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:101]: Why not use single quotes?
Also formatting wise, should be a space between name and id

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 3]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+              String userStmt =
+                  "INSERT INTO users (name,email) VALUES ('" + null + "', '" + email + "');";
+              addStatementToDataBase(userStmt, conn);
+              id = getLatestInsertedID(conn);
+            }
+
+            //insert participants to database
+            String participantsStmt = String.format(
+                "INSERT INTO participants (user, competition, amt_available, net_worth, rank, rank_yesterday) VALUES "
+                    + "(%d,%d,%d,%d,%d,%d);",
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:124]: formatting: space between vars

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 4]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+        //add participants to database
+        for (int i = 0; i < participants.length(); i++) {
+          String email = participants.get(i).toString();
+          String findUserStmt = "SELECT name,id FROM users WHERE email=\"" + email + "\";";
+
+          //check if the user is already in the database or not
+          try (PreparedStatement stmt = conn.prepareStatement(findUserStmt)) {
+            ResultSet rs = stmt.executeQuery();
+            String name = null;
+            int id = -1;
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:107]: use long for large dbs

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 5]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+            while (rs.next()) {
+              name = rs.getString(1);
+              id = rs.getInt(2);
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:110]: getLong()

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 6]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String frontEndInfo = getInformationFromFrontEnd(request);
+
+    try {
+      //get specific information from json object
+      JSONObject jsonObj = new JSONObject(frontEndInfo);
+      int userId = jsonObj.getInt("userId");
+      String compName = jsonObj.getString("name");
+      SimpleDateFormat formatter = new SimpleDateFormat("dd-MM-yyyy");
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:81]: what does this look like at frontend? Are we dealing with different date formatting options there? (eg do we know this isn't MM-DD-YYYY?)

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 7]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String frontEndInfo = getInformationFromFrontEnd(request);
+
+    try {
+      //get specific information from json object
+      JSONObject jsonObj = new JSONObject(frontEndInfo);
+      int userId = jsonObj.getInt("userId");
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:79]: Can you error check for this not being an integer? See above in doGet()

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 8]
```diff
@@ -67,24 +69,139 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+          "INSERT INTO competitions (start_date, end_date, competition_name, creator) VALUES "
+              + "(DATE '%tF', DATE '%tF', '%s', %d);",
+          startDate, endDate, compName, userId);
+
+      DataSource pool = (DataSource) request.getServletContext().getAttribute("db-connection-pool");
+      try (Connection conn = pool.getConnection()) {
+
+        //add competition to database
+        addStatementToDataBase(competitionStmt, conn);
+        int competitionId = getLatestInsertedID(conn);
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:96]: Make long

**Review by roonil-wazlib — CHANGES_REQUESTED** [review]

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java (hunk 6 above)
Comment (Phoebe-Khokgawe) [reply project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java]: Yes we do. The data that i get from front end is 'dd-mm-yyyy' in String so I'm only converting what I got in String to a date object. But I can try having a look into converting to date object but not specifying the format

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Phoebe-Khokgawe) [reply project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java]: removed this function

**Comment by Phoebe-Khokgawe** [comment]
@roonil-wazlib all the request changes are fixed here 00f9daacae02e682ed97d836c6a0a13a44734ce5

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 9]
```diff
@@ -83,48 +83,58 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+        }
+        long competitionId = getLatestInsertedID(conn);
 
         //add participants to database
         for (int i = 0; i < participants.length(); i++) {
           String email = participants.get(i).toString();
-          String findUserStmt = "SELECT name,id FROM users WHERE email=\"" + email + "\";";
 
           //check if the user is already in the database or not
+          String findUserStmt = "SELECT name,id FROM users WHERE email='" + email + "';";
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:106]: space after name,

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 10]
```diff
@@ -83,48 +83,58 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-        addStatementToDataBase(competitionStmt, conn);
-        int competitionId = getLatestInsertedID(conn);
+        String competitionStmt = String.format(
+            "INSERT INTO competitions (start_date, end_date, competition_name, creator) VALUES "
+                + "(DATE '%tF', DATE '%tF', '%s', %d);",
+            startDate, endDate, compName, userId);
+        try (PreparedStatement competitionstmt = conn.prepareStatement(competitionStmt)) {
+          // Execute the statement
+          competitionstmt.execute();
+          LOGGER.log(Level.INFO, "Added " + competitionstmt + "to database.");
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:97]: *

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 11]
```diff
@@ -83,48 +83,58 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
             if (name == null) {
-              //add user to database
+              //add a new user to database
               String userStmt =
                   "INSERT INTO users (name,email) VALUES ('" + null + "', '" + email + "');";
-              addStatementToDataBase(userStmt, conn);
+              try (PreparedStatement statement = conn.prepareStatement(userStmt)) {
+                // Execute the statement
+                statement.execute();
+                LOGGER.log(Level.INFO, "Added " + statement + "to database.");
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:123]: *

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 12]
```diff
@@ -83,48 +83,58 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+              }
               id = getLatestInsertedID(conn);
             }
 
             //insert participants to database
             String participantsStmt = String.format(
                 "INSERT INTO participants (user, competition, amt_available, net_worth, rank, rank_yesterday) VALUES "
-                    + "(%d,%d,%d,%d,%d,%d);",
+                    + "(%d , %d , %d , %d , %d , %d);",
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:131]: no spaces before comments

File: project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java [hunk 13]
```diff
@@ -83,48 +83,58 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
                 id, competitionId, 500, 500, 1, null);
-
-            addStatementToDataBase(participantsStmt, conn);
+            try (PreparedStatement statement = conn.prepareStatement(participantsStmt)) {
+              // Execute the statement
+              statement.execute();
+              LOGGER.log(Level.INFO, "Added " + statement + "to database.");
```
Comment (roonil-wazlib) [inline project/src/main/java/com/google/sps/servlets/CompetitionsServlet.java:136]: *

**Review by roonil-wazlib — APPROVED** [review]
LGTM after comments

**Review by texm — APPROVED** [review]

**Pull request merged**